core/src/main/java/feast/core/service/JobCoordinatorService.java

```diff
@@ -143,68 +137,111 @@ void startOrUpdateJobs(List<JobUpdateTask> tasks) {
-      if (sourceToStoreJob.isPresent()) {
-        Job job = sourceToStoreJob.get();
-        // Record the job as required to safeguard it from getting stopped
-        requiredJobs.add(sourceToStoreJob.get());
+
+    for (Map.Entry<Source, Set<Store>> mapping : sourceStoresMappings.entrySet()) {
+      Source source = mapping.getKey();
+      Set<Store> stores = mapping.getValue();
+      Set<FeatureSet> featureSets =
+          stores.stream()
+              .flatMap(s -> getFeatureSetsForStore(s).stream())
+              .collect(Collectors.toSet());
+
+      Job job = getOrCreateJob(source, stores);
+
+      if (job.isDeployed()) {
         if (!job.isRunning()) {
+          jobTasks.add(UpdateJobStatusTask.builder().setJob(job).setJobManager(jobManager).build());
+
           // Mark that it is not safe to stop jobs without disrupting ingestion
           isSafeToStopJobs = false;
+          continue;
+        }
+
+        if (jobRequiresUpgrade(job, stores)) {
+          job.setStores(stores);
+
+          jobTasks.add(UpgradeJobTask.builder().setJob(job).setJobManager(jobManager).build());
+        } else {
+          jobTasks.add(UpdateJobStatusTask.builder().setJob(job).setJobManager(jobManager).build());
         }
+      } else {
+        jobTasks.add(CreateJobTask.builder().setJob(job).setJobManager(jobManager).build());
       }
 
-      jobUpdateTasks.add(
-          new JobUpdateTask(
-              getFeatureSetsForStore(store),
-              source,
-              store,
-              sourceToStoreJob,
-              jobManager,
-              updateTimeout,
-              JobTargetStatus.RUNNING));
+      allocateFeatureSets(job, featureSets);
+
+      // Record the job as required to safeguard it from getting stopped
+      activeJobs.add(job);
     }
-    // Stop extra jobs that are not required to mantain ingestion when safe
+    // Stop extra jobs that are not required to maintain ingestion when safe
     if (isSafeToStopJobs) {
-      getExtraJobs(requiredJobs)
+      getExtraJobs(activeJobs)
           .forEach(
               extraJob -> {
-                jobUpdateTasks.add(
-                    new JobUpdateTask(
-                        extraJob.getFeatureSets(),
-                        extraJob.getSource(),
-                        extraJob.getStore(),
-                        Optional.of(extraJob),
-                        jobManager,
-                        updateTimeout,
-                        JobTargetStatus.ABORTED));
+                jobTasks.add(
+                    TerminateJobTask.builder().setJob(extraJob).setJobManager(jobManager).build());
               });
     }
 
-    return jobUpdateTasks;
+    return jobTasks;
   }
 
-  /** Get the non terminated ingestion job ingesting between source and source if exists. */
-  @Transactional
-  public Optional<Job> getJob(Source source, Store store) {
+  private boolean jobRequiresUpgrade(Job job, Set<Store> stores) {
```

Comment:
For both `jobRequiresUpgrade` and `allocateFeatureSets` it would be great to have a short comment on what they are doing. Does `jobRequiresUpgrade` only care about feature set consistency, or also sinks or stores or other configuration? What does allocation mean?